python/lsst/ts/m2/mock_server.py

```diff
@@ -0,0 +1,924 @@
+            )
+
+        elif name == "cmd_setTemperatureOffset":
+            await self._write_msg_temperature_offset(
+                msg["ring"], msg["intake"], msg["exhaust"]
+            )
+
+        # Command result
+        if self._is_cmd(name):
+            id_success = CommandStatus.Success
+            msg_success = {"id": id_success.name.lower(), "sequence_id": sequence_id}
+            await self._write_json_packet(self.server_cmd.writer, msg_success)
+
+        await asyncio.sleep(period_check)
+
+    def _decode_incoming_msg(self, msg_encode):
+        """Decode the incoming message.
+
+        Parameters
+        ----------
+        msg_encode : `bytes` or `None`
+            Encoded message.
+
+        Returns
+        -------
+        `dict`
+            Decoded messge. If the msg_encode is None or can not be decoded by
+            JSON, return an empty dictionary.
+        """
+
+        # Decode the message
+        if msg_encode is None:
+            return dict()
+        else:
+            msg_decode = msg_encode.decode()
+
+        # Decode by JSON
+        try:
+            return json.loads(msg_decode)
+
+        except json.JSONDecodeError:
+            self.log.debug(f"Can not decode the message: {msg_decode}.")
+            return dict()
+
+    def _is_cmd(self, id_value):
+        """Is the command or not.
+
+        Parameters
+        ----------
+        id_value : `str`
+            ID value.
+
+        Returns
+        -------
+        `bool`
+            True if this is a command.
+        """
+
+        return "cmd_" in id_value
+
+    def _connect_callback_tel(self, server_tel):
+        """Called when the telemetry server connection state changes.
+
+        Parameters
+        ----------
+        server_tel : `tcpip.OneClientServer`
+            Telemetry server.
+        """
+        self._monitor_loop_task_tel.cancel()
+
+        if self.server_tel.connected:
+            self._monitor_loop_task_tel = asyncio.create_task(self._monitor_msg_tel())
+
+    async def _monitor_msg_tel(self, period_tel=0.05, period_check=0.001):
+        """Monitor the message of incoming telemetry.
+
+        Parameters
+        ----------
+        period_tel : `float`, optional
+            Telemetry period in second. The frequency is 20 Hz usually. (the
+            default is 0.05)
+        period_check : `float`, optional
+            Period of checking the incoming messages in second. (the default is
+            0.001)
+        """
+
+        try:
+            while True:
+
+                await self._write_msg_tel(period_tel)
+
+                if self.server_tel.reader.at_eof():
+                    self.log.info("Telemetry reader at eof; closing client")
+                    break
+
+                await self._process_msg_tel(period_check)
+
+        except ConnectionError:
+            self.log.info("Telemetry reader disconnected; closing client")
+            self._monitor_loop_task_tel.cancel()
```

Comment:
As I replied above, I found this might be needed when I used the `iPython`. The following line of `await self.server_tel.close_client()` is executed in my test. Thanks!